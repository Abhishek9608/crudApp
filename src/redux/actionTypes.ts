export enum ActionTypes {
  ADD_POST = "ADD_POST",
  DELETE_POST = "DELETE_POST",
  UPDATE_POST = "UPDATE_POST",
  GET_POSTS = "GET_POSTS"
}
