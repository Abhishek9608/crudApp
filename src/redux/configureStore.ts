import thunkMiddleware from "redux-thunk";

import {
  createStore, 
  applyMiddleware, 
} from "redux";

import reducers from "./reducer";

export const configureStore = () => {
  const store = createStore(
    reducers,
    applyMiddleware(thunkMiddleware)
  );

  return store;
};
