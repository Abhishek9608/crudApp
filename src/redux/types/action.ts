import { post } from "../../types";
import { ActionTypes } from "../actionTypes";

interface AddPost {
  type: ActionTypes.ADD_POST;
  todoDetails: { id: string; description: string };
}
interface DeletePost {
  type: ActionTypes.DELETE_POST;
  todoId: string;
}
interface UpdatePost {
  type: ActionTypes.UPDATE_POST;
  todoDetails: { id: string; description: string };
}

interface GetPosts {
  type: ActionTypes.GET_POSTS;
  posts: Array<post>
}

export type PostsAction = 
AddPost | 
DeletePost | 
UpdatePost |
GetPosts;
