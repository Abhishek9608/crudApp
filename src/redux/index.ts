import { ThunkAction } from "redux-thunk";
import { configureStore } from "./configureStore";

// Redux store
export const store = configureStore();

export type RootState = ReturnType<typeof store.getState>;
export type Dispatch = typeof store.dispatch;
export type ActionType = ReturnType<typeof store.dispatch>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  ActionType
>;
