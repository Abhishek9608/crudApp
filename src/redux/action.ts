import { AppThunk } from "./index";
import {ActionTypes} from './actionTypes'
import { data } from "../types";

export const fetchPosts = () : AppThunk  =>{
  return (dispatch, getState) => {
    fetch("https://jsonplaceholder.typicode.com/posts")
      .then((response) => response.json())
      .then((json) => {
        console.log(json)
        dispatch({
          type: ActionTypes.GET_POSTS,
          posts: json
        })
      });
  };
}

export const addPost = (data : data): AppThunk => {
  return (dispatch) => {
    fetch('https://jsonplaceholder.typicode.com/posts', {
    method: 'POST',
    body: JSON.stringify(data)
  })
    .then((response) => response.json())
    .then((json) =>{
      dispatch(fetchPosts())
    });
  }
}

export const editPost = (data : data, id:number | undefined): AppThunk => {
  return (dispatch) => {
    fetch(`https://jsonplaceholder.typicode.com/posts/${id}`, {
    method: 'PUT',
    body: JSON.stringify(data)
  })
    .then((response) => response.json())
    .then((json) =>{
      dispatch(fetchPosts())
    });
  }
}


export const deletePost = (id : number): AppThunk => {
  return (dispatch) => {
    fetch(`https://jsonplaceholder.typicode.com/posts/${id}`, {
    method: 'DELETE',
    // body: JSON.stringify(id)
  })
    .then((response) => response.json())
    .then((json) =>{
      dispatch(fetchPosts())
    });
  }
}
