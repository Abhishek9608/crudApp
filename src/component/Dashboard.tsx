import  React, {useEffect, useState} from 'react';
import Card from '@mui/material/Card';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';
import AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import IconButton from '@mui/material/IconButton';
import FormGroup from '@mui/material/FormGroup';
import PostModal from './PostModal';
import {useAppDispatch, useAppSelector} from '../redux/types/hooks'
import {addPost, deletePost, editPost, fetchPosts } from '../redux/action'
import { data, post } from '../types';

const  Dashboard = ()  => {
  const [modalType, setModalType] = useState('')
  const [open, setOpen] = React.useState(false);
  const [selectedCard, setSelectedCard] = useState<post | null>(null)


  const posts = useAppSelector(state => state.posts)
  const dispatch = useAppDispatch();

  const handleOpen = () => setOpen(true);
  const handleClose = () => setOpen(false);


  useEffect(() => {
    dispatch(fetchPosts())
  }, [])


  const handleAdd = (data:data) => {
    dispatch(addPost(data))
    handleClose()
  }

  const handleEdit = (data:data) => {
    dispatch(editPost(data, selectedCard?.id))
    handleClose()
  }

  const handleDelete = (id: number) => {
    dispatch(deletePost(id))
  }
 
  const handleType =(type: string, data:post | null) => {
    if(type == 'edit'){
      setSelectedCard(data)
    }
    setModalType(type);
    handleOpen()
  }

  return (
    <div>
      <Box sx={{ flexGrow: 1 }}>
      <FormGroup>
      </FormGroup>
      <AppBar position="static">
        <Toolbar>
          <IconButton
            size="large"
            edge="start"
            color="inherit"
            aria-label="menu"
            sx={{ mr: 2 }}
          >
          </IconButton>
          <Typography variant="h6" component="div" sx={{ flexGrow: 1 }}>
            BLOGS
          </Typography>
            <div>
              <IconButton
                size="large"
                aria-label="account of current user"
                aria-controls="menu-appbar"
                aria-haspopup="true"
                color="inherit"
              >
              </IconButton>
            </div>
        </Toolbar>
      </AppBar>
      </Box>

      <div style={{marginTop:"20px"}} >
      <Button variant="contained" onClick={() => handleType('add', null)}>Add Post</Button>
      </div>
        {
          posts?.slice(0 , 10)?.map((item) => (
            <Card style={{margin:"50px"}} >
              <CardContent>
                <Typography gutterBottom variant="h5" component="div">
                  {item.title}
                </Typography>
                <Typography variant="body2" color="text.secondary">
                  {item.body}
                </Typography>
              </CardContent>
              <CardActions>
                
              <Button variant="contained" onClick={() => handleType('edit', item)}>Edit</Button>
              {/* <PostModal buttonText={"Edit"} /> */}
                <Button variant="outlined" onClick={() => handleDelete(item.id)}>
                  Delete
                </Button>
              </CardActions>
            </Card>
          ))
        }
        <PostModal 
          open={open}
          handleClose={handleClose}
          handleClick={modalType === 'add' ? handleAdd : handleEdit}
          selectedCard={modalType === 'edit' ? selectedCard : null}
        />
    </div>
  );
}

export default Dashboard;