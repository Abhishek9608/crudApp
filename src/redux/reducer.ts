import { PostsAction } from "./types/action";
import { ActionTypes } from "./actionTypes";
import { post } from "../types";

interface stateType {
  posts: Array<post>
}

const initialState: stateType = {
  posts: []
};


function TodosReducer(
  state: Readonly<stateType> = initialState,
  action: PostsAction
): stateType {
  switch (action.type) {
    case ActionTypes.GET_POSTS:
      return {
        ...state,
        posts: action.posts
      };

    default:
      return state;
  }
}

export default TodosReducer;
