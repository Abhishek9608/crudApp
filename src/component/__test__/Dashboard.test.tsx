import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Dashboard from '../Dashboard'
import {store} from '../../redux'
import { findByRole, fireEvent, getAllByText, getByText, queryByRole, queryByText, render, waitFor } from '@testing-library/react';
import {deletePost, addPost } from '../../redux/action'


jest.mock('../../redux/action', () => ({
    ...jest.requireActual('../../redux/action'),
    deletePost: jest.fn(),
    addPost: jest.fn()
}))

jest.mock('../../redux/types/hooks', () => ({
    ...jest.requireActual('../../redux/types/hooks'),
    useAppSelector: () => [{
        "userId": 1,
        "id": 1,
        "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
        "body": "quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto"
    }],
    useAppDispatch: () => jest.fn()
}))

describe('Dashboard page', () => {

    test('Dashboard page snapshot should match correctly', () => {
        const tree = renderer
        .create(<Provider store={store}>
            <Dashboard/>
        </Provider>)
        .toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('page title shoulld be there',() => {
        const {queryByText} = render(<Provider store={store}>
            <Dashboard/>
        </Provider>)

        const getHeading = queryByText('BLOGS');
        expect(getHeading).toBeTruthy()
    })

    test('After Clicking on Add post modal should open', () => {
        const {getByRole, getByText} = render(<Provider store={store}>
            <Dashboard/>
        </Provider>)
        const getAddButton = getByRole('button', {name: /add post/i});
        fireEvent.click(getAddButton);
        const Submit = getByRole('button', {name: /submit/i});
        expect(Submit).toBeTruthy()
    })

    test('After Clicking Edit button modal should open', async () => {
        const {getByRole,getByText} = render(<Provider store={store}>
            <Dashboard/>
        </Provider>)
        const getAddButton = getByRole('button', {name: /edit/i});
        fireEvent.click(getAddButton);
        const Submit = getByRole('button', {name: /submit/i});
        expect(Submit).toBeTruthy()
    })
    
    test('After Clicking on delete button, delete function should be called',() => {
        const {getAllByText} = render(<Provider store={store}>
            <Dashboard/>
        </Provider>)

        const getDeletedButton = getAllByText('Delete')[0];
        fireEvent.click(getDeletedButton);
        expect(deletePost).toBeCalled()
    })

    test('After clicking on add, then add API function should be called', async () => {
        const {getByRole, getByTestId, findByRole} = render(<Provider store={store}>
            <Dashboard/>
        </Provider>)
        const getAddButton = getByRole('button', {name: /add post/i});
        fireEvent.click(getAddButton);
        const input = getByTestId('title')
        fireEvent.change(input, { target: { value: 'test' } })
        await waitFor(() => expect(input).toHaveValue("test"));
        const body = getByTestId('title')
        fireEvent.change(body, { target: { value: 'body' } })
        await waitFor(() => expect(body).toHaveValue("body"));
        const Submit = getByRole('button', {name: /submit/i});
        fireEvent.click(Submit);
        expect(addPost).toBeCalled()
    })
})
