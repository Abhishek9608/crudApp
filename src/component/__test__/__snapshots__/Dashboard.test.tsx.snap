// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard page Dashboard page snapshot should match correctly 1`] = `
<div>
  <div
    className="MuiBox-root css-i9gxme"
  >
    <div
      className="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
    />
    <header
      className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
    >
      <div
        className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      >
        <button
          aria-label="menu"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-ancrnh-MuiButtonBase-root-MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        />
        <div
          className="MuiTypography-root MuiTypography-h6 css-1juivf6-MuiTypography-root"
        >
          BLOGS
        </div>
        <div>
          <button
            aria-controls="menu-appbar"
            aria-haspopup="true"
            aria-label="account of current user"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          />
        </div>
      </div>
    </header>
  </div>
  <div
    style={
      Object {
        "marginTop": "20px",
      }
    }
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Add Post
    </button>
  </div>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    style={
      Object {
        "margin": "50px",
      }
    }
  >
    <div
      className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
      >
        sunt aut facere repellat provident occaecati excepturi optio reprehenderit
      </div>
      <p
        className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
      >
        quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto
      </p>
    </div>
    <div
      className="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Edit
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
  <div
    role="modal"
  />
</div>
`;
